### ISSUE TITLE
ssPATHS

### ISSUE THREAD

**Issue by nrosed** [Bioconductor/Contributions#1194]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/nrosed/ssPATHS

Confirm the following by editing each check box to '[x]'

- [ x ] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [ x ] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [ x ] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [ x ] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [ x ] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [ x ] The 'devel' branch for new packages and features.
- [ x ] The stable 'release' branch, made available every six
      months, for bug fixes.
- [ x ] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by nrosed** [comment]
@mtmorgan I have changed the repository of the code. 
It should be  https://github.com/nrosed/ssPATHS
previously it was  https://github.com/ratschlab/ssPATHS

**Comment by lshep** [comment]
I have made the necessary change in our internal database for this issue.  Please ensure your [webhook](https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#adding-a-web-hook) is set up to ensure new builds on version bumps.  Please also correct the ERRORs in the build report before @hpages  does a through review of your package.  Thank you.

**Comment by nrosed** [comment]
@hpages The code is ready for review. Thanks in advance.

**Comment by hpages** [comment]
Hi @nrosed ,

Thanks for submitting ssPATHS. The package is small and quite simple but I would suggest that you use a SummarizedExperiment object as the central container for your expression data. Also the package requires some cleaning. See details below. Thanks!

H.

1. Please familiarize yourself with the way Bioconductor represents expression data. The standard Bioconductor data structure for this is the SummarizedExperiment container. For example, here is how to import your `tcga_expr_df` dataset in a SummarizedExperiment object:
    ```
    library(SummarizedExperiment)
    tcga_se <- SummarizedExperiment(t(tcga_expr_df[ , -(1:4)]), colData=tcga_expr_df[ , 2:4])
    colnames(tcga_se) <- tcga_expr_df$tcga_id
    ```
    Note that the Bioconductor standard is to have 1 row per gene and 1 column per sample. Samples are annotated in the `colData` component. The advantage of this is to keep the metadata separated from the expression data.
    Then your typical workflow should operate on this object. For example formatting the data becomes:
    ```
    hypoxia_gene_ids <- get_hypoxia_genes()
    hypoxia_gene_ids <- intersect(hypoxia_gene_ids, rownames(tcga_se))
    hypoxia_se <- tcga_se[hypoxia_gene_ids, ]  # instead of your hypoxia_df object
    colData(hypoxia_se)$Y <- ifelse(colData(hypoxia_se)$is_normal, 0, 1)
    ```
    Finally the user should be able to call `get_gene_weights()` on `hypoxia_se`. The function will get the sample ids with `colnames(se)` (assuming `se` is the name of the argument of the function), it will get `Y` with `colData(se)$Y`, and it will get the expression matrix with `assay(se)`.

2. The following files in your `vignettes/` folder are intermediate products of the vignette building process: `*.pdf`, `*.tex`, and `ssPATHS.synctex.gz`. These files should not be added to your git repository. Please remove them. The only file that you need to keep here is `ssPATHS.Rnw`.

3. Looks like you only use the ssPATHS, ROCR, and ggplot2 packages in your vignette so only load these packages (this is in the code chunk following "First, we load the appropriate packages we will need").

4. The following `library()` statements in your `R/ssPATHS.R` file:
    ```
    library("dml")
    library("data.table")
    library("ROCR")
    library("MESS")
    ```
    are not needed because you are importing these packages. Please remove them (they are actually considered bad practice).

5. Please only load the packages you need in your examples. For example you don't seem to use the dml, data.table, or MESS package anywhere in your examples so there is no reason to start the examples with all these `library()` statements. Note that only some of your examples need the ROCR package so load it only when needed (e.g. the examples for `get_hypoxia_genes()` don't need it).

6. Same problem with your unit tests. None of the `library()` statements you have in them is needed. Please remove.

7. You don't seem to use the data.table package at all so you should remove it from the `Imports` field (`DESCRIPTION` file) and from the `NAMESPACE` file.

8. I don't think you use the plyr package at all either so you should remove it from the `Suggests` field.

9. What is the purpose of the `print(getwd())` statement in the examples for `get_gene_weights()`?

10. Please clean your `*.Rd` files. They contain a lot of comments (i.e. lines that start with `%`) that belong to the original man page stubs generated by `package.skeleton()`. These comments should be removed. The empty sections should also be removed (`R CMD check` emits many notes about these empty sections). And the `\keyword` lines at the bottom of the man pages need to be fixed.

11. In various places where you have a data.frame you describe it as a matrix. For example the man page for `get_gene_weights()` says that `expression_matr` should be an expression matrix when the function in fact expects a data.frame. Note that the reason `expression_matr` cannot be a matrix is because the data is heterogeneous i.e. the columns are of mixed types (some contain numeric values and some contain strings). Note that this issue will go away when you switch to using a SummarizedExperiment object instead of a data.frame or matrix for storing the expression data.

12. Typo in the caption of the last figure of the vignette: "we find the the"

13. Coding style:

    - Use `<-` instead of `=` for assignments.

    - Note that setting normal samples to 0 and tumor samples to 1 can be done simply with:
        ```
        hypoxia_df$Y <- ifelse(tcga_expr_df$is_normal, 0, 1)
        ```
        instead of the 3 lines:
        ```
        hypoxia_df$Y = 0
        hypoxia_df$Y[tcga_expr_df$is_normal==TRUE] = 0
        hypoxia_df$Y[tcga_expr_df$is_normal==FALSE] = 1
        ````

**Comment by nrosed** [comment]
Hi @hpages , 
I believe I addressed all of your comments and changed everything you requested. Thank you for the detailed feedback, and I hope you find the package in better shape now.

**Comment by hpages** [comment]
Hi Nathalie,

Thanks for all the changes.

- You use SummarizedExpression instead of SummarizedExperiment in several places. There is no such thing as SummarizedExpression objects.

- You should put SummarizedExperiment in the Depends field instead of the Imports field. That way you never need to load it explicitly with `library(SummarizedExperiment)`.

- My point about using SummarizedExperiment objects to represent gene expression values was that it's a better representation than the data.frame representation. So you should make SummarizedExperiment your central container and do all your manipulations on this. Right now in your man pages you sometimes operate on `tcga_expr_df` and sometimes you turn it into a SummarizedExperiment object and operate on that object. Instead, the first thing you should do with `tcga_expr_df` is turn it into a SummarizedExperiment object. The end user should never have to deal with `tcga_expr_df`. You could add a little utility function, say `make_SummarizedExperiment_from_TCGA()`, and use call it on `tcga_expr_df` or `new_samp_df` right after doing `load(tcga_expr_df)` or `load(new_samp_df)` in your man page examples or vignette. Or, alternatively, you could serialize these SummarizedExperiment objects to replace the serialized data.frames.

In other words, you need to fully embrace the use of SummarizedExperiment objects.

Thanks,
H.

**Comment by nrosed** [comment]
Hi @hpages,

1. I removed all mentions of SummarizedExpression.
2. I moved SummarizedExperiment to the Depends field
3. I looked through all the man pages and now convert the input data from a data.frame to a SummarizedExperiment object right after it is loaded. I don't think I operate on `tcga_expr_df` beyond converting it to a SummarizedExperiment object anymore. 

Thanks!

**Comment by hpages** [comment]
Hi Nathalie,

Thanks again for the changes. Only a few small leftovers:

14. As mentioned previously, the main benefit of moving SummarizedExperiment from Imports to Depends is that when the user loads ssPATHS with `library(ssPATHS)` this automatically makes SummarizedExperiment available. So you don't need to load SummarizedExperiment explicitly with `library(SummarizedExperiment)`. Please remove these statements from you man pages and vignette.

15. In the man pages for `get_gene_weights`, `get_classification_accuracy`, and `get_new_samp_score`, the following line is not needed and confusing because `colData(hypoxia_se)` already has the correct `sample_id` column:
    ```
    colData(hypoxia_se)$sample_id <- colData(tcga_se)$sample_id
    ```

16. When constructing `new_samp_se` in the man page for `get_new_samp_score`, it's preferable to do:
    ```
    new_samp_se <- SummarizedExperiment(t(new_samp_df[ , -(1)]), colData=new_samp_df[ , 1, drop=FALSE])
    ```
    instead of:
    ```
    new_samp_se <- SummarizedExperiment(t(new_samp_df[ , -(1)]), colData=new_samp_df[ , 1])
    colnames(colData(new_samp_se)) <- "sample_id"
    ```

17. Please don't use lines of code that exceed 80 characters in your man pages (they display poorly when opening the man page in a terminal). So for example, instead of:
    ```
    tcga_se <- SummarizedExperiment(t(tcga_expr_df[ , -(1:4)]), colData=tcga_expr_df[ , 2:4])
    ````
    do:
    ```
    tcga_se <- SummarizedExperiment(t(tcga_expr_df[ , -(1:4)]),
                                    colData=tcga_expr_df[ , 2:4])
    ```
    (Note the indentation of the 2nd line with respect to the previous line.)

18. In various man pages you have:

    > The colData object must contain Y (whether this is a positive or negative sample), sample_id (unique sample id).

    However it would be much clearer to say something like:

    > The colData component must contain columns \code{Y} and \code{sample_id}. The former indicates whether this is a positive or negative sample and the latter is the unique sample id.

19. The `gene_ids` column of the `gene_weights_reference` dataset should be `gene_id` (singular) for consistency with the name of the other column (`gene_weight`), and also for consistency with the other datasets in your package.
 
We're almost there.
H.

**Comment by nrosed** [comment]
Hi @hpages, 

Thanks again for your detailed feedback, hopefully now everything is in order.

-Natalie

**Comment by hpages** [comment]
Hi Nathalie,

Please pay attention to item 16. The whole purpose of the change I suggested was to avoid having to do the ugly `colnames(colData(new_samp_se)) <- "sample_id"`. Not a deal breaker. I'm accepting the package. Hopefully you'll be able to address this very minor issue before the release.

Note that now that the package is accepted it will soon be added to our git server at git.bioconductor.org so you'll have to remember to push all your changes to this location in addition to pushing them to GitHub (you will soon receive detailed instructions about this).

Cheers,
H.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/nrosed.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("ssPATHS")`. The package 'landing page' will be created at

https://bioconductor.org/packages/ssPATHS

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
